include/metal/list/sort.hpp

```diff
@@ -75,35 +77,116 @@ namespace metal {
         template<
             typename, typename, typename, template<typename...> class,
             typename = true_>
-        struct _merge {};
+        struct _partial_merge {};
 
         template<
             typename x, typename y, typename... zs,
             template<typename...> class e>
-        struct _merge<
+        struct _partial_merge<
             x, y, list<zs...>, e, if_<call<e, head<y>, head<x>>, true_, false_>>
-            : _merge<x, tail<y>, list<zs..., head<y>>, e> {};
+            : _partial_merge<x, tail<y>, list<zs..., head<y>>, e> {};
 
         template<
             typename x, typename y, typename... zs,
             template<typename...> class e>
-        struct _merge<
+        struct _partial_merge<
             x, y, list<zs...>, e, if_<call<e, head<y>, head<x>>, false_, true_>>
-            : _merge<tail<x>, y, list<zs..., head<x>>, e> {};
+            : _partial_merge<tail<x>, y, list<zs..., head<x>>, e> {};
 
         template<typename... xs, typename... zs, template<typename...> class e>
-        struct _merge<list<xs...>, list<>, list<zs...>, e> {
-            using type = list<zs..., xs...>;
+        struct _partial_merge<list<xs...>, list<>, list<zs...>, e> {
+            using type = list<zs...>;
+            using l1_carry = list<xs...>;
+            using l2_carry = list<>;
         };
 
         template<typename... ys, typename... zs, template<typename...> class e>
-        struct _merge<list<>, list<ys...>, list<zs...>, e> {
-            using type = list<zs..., ys...>;
+        struct _partial_merge<list<>, list<ys...>, list<zs...>, e> {
+            using type = list<zs...>;
+            using l1_carry = list<>;
+            using l2_carry = list<ys...>;
         };
 
         template<typename z, template<typename...> class e>
-        struct _merge<list<>, list<>, z, e> {
+        struct _partial_merge<list<>, list<>, z, e> {
             using type = z;
+            using l1_carry = list<>;
+            using l2_carry = list<>;
+        };
+
+        template<class, class, typename = true_>
+        struct _large_partial_merge {};
+
+        template<class... e1s, class... e2s>
+        struct _large_partial_merge<
+            list<e1s...>, list<e2s...>,
+            number<(
+                (sizeof...(e1s) >= 2) && (sizeof...(e2s) >= 2)
+                && (sizeof...(e1s) + sizeof...(e2s)) > 512)>> {
```

Comment:
To keep things more readable, I suggest using `metal::size` and friends here, so that we don't need to pattern match the variadic pack, nor use `sizeof...` everywhere.